File: 65-valid-number/65-valid-number.py
class Solution(object):
    def isNumber(self, s):
        """
        :type s: str
        :rtype: bool
        """
        def is_integer(s,ind,right):
            if s == '' or not s:
                return True
            if ind == 1 and len(s) == 1 and s[0] in ['+','-']:
                return False
            if ind == 0 and len(s) == 1 and s[0] in ['+','-'] and (not right or (right[0] in ['+','-'] and len(right) == 1)):
                return False
            if not ('0' <= s[0] <= '9') and not s[0] in ['+','-']:
                return False
            i = 0
            if s[0] in ['+','-']:
                i = 1
            n = len(s)
            while(i < n):
                if not ('0' <= s[i] <= '9'):
                    return False
                i += 1
            return True
        def is_decimal(s):
            if s == '' or not s:
                return True
            if not ('0' <= s[0] <= '9') and s[0] not in ['+','-'] and s[0] != '.':
                return False
            i = 0
            n = len(s)
            integer = []
            integer1 = []
            flag = True
            while(i < n):
                if s[i] == '.' and flag:
                    flag = False
                    integer1[:] = integer[:]
                    integer = []
                elif s[i] == '.' and not flag:
                    return False
                elif i > 0 and not ('0' <= s[i] <= '9'):
                    return False
                else:
                    integer.append(s[i])
                i = i + 1
            if not integer and not integer1:
                return False
            elif not flag and not is_integer(integer1,0,integer):
                return False
            elif not flag and is_integer(integer,1,[]) and is_integer(integer1,0,integer):
                return True
            elif flag and is_integer(integer,0,[]):
                return True
            return False
        def is_valid(s):
            if s == '' or not s:
                return False
            if not ('0' <= s[0] <= '9') and s[0] not in ['+','-'] and s[0] != '.':
                return False
            decimal, decimal1 = [], []
            i = 0
            n = len(s)
            flag = True
            while(i < n):
                if s[i] in ['e','E'] and flag:
                    flag = False
                    decimal1[:] = decimal[:]
                    decimal = []
                    j = i
                elif s[i] in ['e','E'] and not flag:
                    return False
                elif i > 0 and not flag and i != (j+1) and not ('0' <= s[i] <= '9'):
                    return False
                else:
                    decimal.append(s[i])
                i = i + 1
            if (not decimal or not decimal1) and not flag:
                return False
            elif not flag and is_integer(decimal,1,[]) and is_decimal(decimal1):
                return True
            elif flag and is_decimal(decimal):
                return True
            else:
                return False
        return is_valid(s)
                    
                          
                    
                    

File: 65-valid-number/65-valid-number.java
class Solution {
    public boolean isNumber(String s) {
        //find the e or E to extract the number
        int i = 0;
        int sLen = s.length();
        while(i < sLen){
            char curChar = s.charAt(i);
            if(curChar == 'e' || curChar == 'E')break;
            i++;
        }
        //no e or E
        String decimal = s.substring(0, i);
        if(i == sLen)return isDecimal(decimal);
        String integer = s.substring(i+1);
        return isDecimal(decimal) && isInteger(integer);
        
    }
    
    public boolean isDecimal(String s){
        //extract two parts integers
        int sLen = s.length();
        if(sLen == 0)return false;
        
        //exclude the first sign digit
        int i = 0;
        int start = 0;
        //exclude the first sign digit
        if(s.charAt(0) == '+' || s.charAt(0) == '-'){
            i += 1;
            start+=1;
        }
        while(i < sLen){
            char curChar = s.charAt(i);
            if(curChar == '.')break;
            i++;
        }
        String secondInteger = "";
        String firstInteger = s.substring(start, i);
        //have '.'
        if(i != sLen)secondInteger = s.substring(i+1);
        boolean res = true;
        if(firstInteger.length() != 0){
            if(firstInteger.charAt(0) == '+' || firstInteger.charAt(0) == '-')return false;
            res &= isInteger(firstInteger);
        }
        if(secondInteger.length() != 0){
            if(secondInteger.charAt(0) == '+' || secondInteger.charAt(0) == '-')return false;
            res &= isInteger(secondInteger);
        }
        if(firstInteger.length() == 0 && secondInteger.length() == 0)return false;
        return res;
    }
    
    
    
    
    public boolean isInteger(String s){
        int sLen = s.length();
        boolean isSign = false;
        
        if (sLen == 0)return false;
        
        if(sLen == 1 && (s.charAt(0) == '+' || s.charAt(0) == '-'))return false;
        int i = 0;
        
        //exclude the first sign digit
        if(s.charAt(0) == '+' || s.charAt(0) == '-'){
            i += 1;
        }
        
        //check if all characters are digits
        while(i < sLen){
            char curChar = s.charAt(i);
            if(!(curChar <= '9' && curChar >= '0'))return false;
            i++;
        }
        return true;
    }
    
    
}